Add race detection to e2e tests

charlesakalugwu opened the pull request:
This adds race detection to our unit and e2e tests

/cc @jim-minter @grantomation @Makdaam 

```release-note
NONE
```

charlesakalugwu commented:
e2e-create flake: #1628 #1643
unit failure: test timed out, adding `-timeout` flag to `go test`

charlesakalugwu commented:
upgrade-v3.2 flake #1410 
/retest

jim-minter commented:
Why is a unit test taking 10 minutes to run?  Unit tests should be quick.

asalkeld commented:
> Why is a unit test taking 10 minutes to run? Unit tests should be quick.

1. "-race" makes the tests a lot slower. 
2. we already have coverage on, which makes it slower
3. pkg/config and pkg/plugin (integration tests) are not super fast ATM
```
go test ./pkg/config/... -coverprofile=coverage.out -covermode=atomic
ok  	github.com/openshift/openshift-azure/pkg/config	0.004s	coverage: 0.0% of statements
ok  	github.com/openshift/openshift-azure/pkg/config/v3	7.410s	coverage: 80.2% of statements
ok  	github.com/openshift/openshift-azure/pkg/config/v4	8.248s	coverage: 80.0% of statements
ok  	github.com/openshift/openshift-azure/pkg/config/v5	7.977s	coverage: 80.0% of statements
ok  	github.com/openshift/openshift-azure/pkg/config/v6	8.457s	coverage: 80.0% of statements
$ go test ./pkg/config/... -coverprofile=coverage.out -covermode=atomic -race
ok  	github.com/openshift/openshift-azure/pkg/config	1.014s	coverage: 0.0% of statements
ok  	github.com/openshift/openshift-azure/pkg/config/v3	32.051s	coverage: 80.2% of statements
ok  	github.com/openshift/openshift-azure/pkg/config/v4	33.897s	coverage: 80.0% of statements
ok  	github.com/openshift/openshift-azure/pkg/config/v5	32.549s	coverage: 80.0% of statements
ok  	github.com/openshift/openshift-azure/pkg/config/v6	32.731s	coverage: 80.0% of statements

$ go test ./pkg/plugin/... -coverprofile=coverage.out -covermode=atomic
ok  	github.com/openshift/openshift-azure/pkg/plugin	50.446s	coverage: 73.6% of statements
$ go test ./pkg/plugin/... -coverprofile=coverage.out -covermode=atomic -race
ok  	github.com/openshift/openshift-azure/pkg/plugin	296.042s	coverage: 73.6% of statements
```

asalkeld commented:
this https://blog.golang.org/race-detector says it can be up to 10x slower. Maybe turn it off for the unit tests? 
"
Because of its design, the race detector can detect race conditions only when they are actually triggered by running code, which means it's important to run race-enabled binaries under realistic workloads.
"

ehashman commented on Makefile:
```diff
@@ -70,7 +70,7 @@ verify:
 
 unit: generate
-	go test ./... -coverprofile=coverage.out -covermode=atomic
+	go test ./... -coverprofile=coverage.out -covermode=atomic -race -timeout 20m
```
A 20m timeout is really long for unit tests! Perhaps we should only include the race detection in long-running CI/e2e tests.

charlesakalugwu replied:
Yes I agree :) The tests in `pkg/plugin` are just monsters and take more time with `-race` enabled and even more time when all of that is run on those highly-cpu-constrained containers in CI.

[This commit](https://github.com/openshift/openshift-azure/pull/1687/commits/df8fc5085bd64a14dfbbf103eaa174c531d87160) already improves the situation significantly notably when running with multiple cores. While this optimization will probably not make much of a difference in CI, I imagine that such optimizations could increase the speed of our sync reconciliation on clusters and could lead to standing up a healthy cluster in less time.

charlesakalugwu commented:
> Why is a unit test taking 10 minutes to run? Unit tests should be quick.

@jim-minter @asalkeld @ehashman as you all point out, the tests in `pkg/plugin` take too long to run with `-race` enabled. I think the issue is that the tests in that package are slow anyway due to certain slow parts of the code.

I tracked down one such problem in `sync.readDB` in `pkg/sync`. I've added a commit which uses an error-handling waitgroup to drastically improve performance of the function (by about 100 seconds). 

```
[charles@akalugwu openshift-azure]$  go test ./pkg/plugin/... -coverprofile=coverage.out -covermode=atomic -race
ok      github.com/openshift/openshift-azure/pkg/plugin 191.966s        coverage: 73.6% of statements
```

Another area which is slow along the codepaths in those tests is the  `*Upgrade.HealthCheck` method in `pkg/cluster`. I will have a deeper look and try to optimize that area. 

I don't think we should leave `-race` out of unit tests because that helps to catch simple data races when we write concurrent code. My first attempt at making `sync.readDB` more concurrent involved using a classic waitgroup with channels and `-race` helped me find and fix a simple data race issue which would otherwise have gone unnoticed.

thekad commented:
> I don't think we should leave `-race` out of unit tests because that helps to catch simple data races when we write concurrent code. My first attempt at making `sync.readDB` more concurrent involved using a classic waitgroup with channels and `-race` helped me find and fix a simple data race issue which would otherwise might have gone unnoticed.

This feels more like a solution waiting for a problem to me? `sync.readDB` wasn't concurrent to begin with. I also think adding -race to unit tests is a bit too much. Definitely useful for e2e and other tests though.

my 2 cents.

charlesakalugwu commented:
@jim-minter @asalkeld @ehashman @thekad I have taken out `-race` for unit tests.

kwoodson commented:
This lgtm.

thekad commented:
LGTM as well, but I'll let one of the more experienced go people to actually slash-lgtm it.

charlesakalugwu commented:
@thekad @kwoodson 

Given the following benchmark test placed in `pkg/sync/v6/sync_test.go`

```
func BenchmarkReadDB(b *testing.B){
	b.ReportAllocs()
	cs := getTestCs()
	for i:= 0; i < b.N; i++ {
		s := &sync{cs: cs}
		s.readDB()
	}
}
```

Here are the results of running the benchmark test before and after the introduction of concurrency

*Before optimization*
```
charles@titan:~/$ git checkout master
M       pkg/sync/v6/sync_test.go
Switched to branch 'master'
Your branch is up to date with 'upstream/master'.

charles@titan:~/$ go test -v -run=BenchmarkReadDB -bench=./pkg/sync/v6/ ./pkg/sync/v6/
PASS
ok      github.com/openshift/openshift-azure/pkg/sync/v6        0.629s

```

*After optimization*
```
charles@titan:~/$ git checkout add-race-detection-to-tests 
M       pkg/sync/v6/sync_test.go
Switched to branch 'add-race-detection-to-tests'
Your branch and 'origin/add-race-detection-to-tests' have diverged,
and have 7 and 2 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)

charles@titan:~/$ go test -v -run=BenchmarkReadDB -bench=./pkg/sync/v6/ ./pkg/sync/v6/
PASS
ok      github.com/openshift/openshift-azure/pkg/sync/v6        0.160s

```

The winner is pretty clear

kwoodson commented:
/lgtm

charlesakalugwu commented:
upgrade-v5.1 flake #1438 
/retest

charlesakalugwu commented:
upgrade-v5.1 flake #1410 
/retest
